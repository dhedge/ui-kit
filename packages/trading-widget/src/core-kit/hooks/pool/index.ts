export { useCheckWhitelist } from './use-check-whitelist'
export { useIsDhedgePool } from './use-is-dhedge-pool'
export { useManagerLogicAddress } from './use-manager-logic-address'
export { useContractPoolComposition } from './use-contract-pool-composition'
export { usePoolComposition } from './use-pool-composition'
export { usePoolDynamicContractData } from './use-pool-dynamic-contract-data'
export { usePoolManagerLogicData } from './use-pool-manager-logic-data'
export { usePoolCompositionWithFraction } from './use-pool-composition-with-fraction'
export { usePoolTokenPrice } from './use-pool-token-price'
export { usePoolFees } from './use-pool-fees'
export { useSynthetixV3AssetBalance } from './synthetixV3/use-synthetix-v3-asset-balance'
export { useTotalFundValueMutable } from './synthetixV3/use-total-funds-value-mutable'
export { useInvalidatePoolContractData } from './use-invalidate-pool-contract-data'
export { usePoolsDynamic } from './multicall'
export { useVaultVestedPoints } from './use-vault-vested-points'
export { useHasDhedgeVaultInComposition } from './use-has-dhedge-vault-in-composition'
