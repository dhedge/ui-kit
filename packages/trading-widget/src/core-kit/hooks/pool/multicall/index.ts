export { usePoolStatic } from './use-pool.static'
export { usePoolManagerStatic } from './use-pool-manager.static'
export { usePoolManagerDynamic } from './use-pool-manager.dynamic'
export { usePoolsDynamic } from './use-pools.dynamic'
