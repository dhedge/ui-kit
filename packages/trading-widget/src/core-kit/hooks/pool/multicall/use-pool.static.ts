import { EasySwapperV2Abi, PoolFactoryAbi, PoolLogicAbi } from 'core-kit/abi'
import { AddressZero } from 'core-kit/const'
import {
  useContractReadErrorLogging,
  useReadContracts,
} from 'core-kit/hooks/web3'
import type {
  MulticallReturnType,
  PoolContractCallParams,
} from 'core-kit/types'
import { getContractAddressById, isZeroAddress } from 'core-kit/utils'

const getContracts = ({ address, chainId }: PoolContractCallParams) =>
  [
    {
      address: getContractAddressById('factory', chainId),
      abi: PoolFactoryAbi,
      functionName: 'isPool',
      chainId,
      args: [address],
    },
    {
      address: address ?? AddressZero,
      abi: PoolLogicAbi,
      functionName: 'poolManagerLogic',
      chainId,
    },
    {
      address: getContractAddressById('easySwapperV2', chainId),
      abi: EasySwapperV2Abi,
      functionName: 'customCooldownDepositsWhitelist',
      args: [address],
      chainId,
    },
    {
      address: getContractAddressById('easySwapperV2', chainId),
      abi: EasySwapperV2Abi,
      functionName: 'customCooldown',
      args: [],
      chainId,
    },
  ] as const

type Data = MulticallReturnType<ReturnType<typeof getContracts>>

const selector = (data: Data) => ({
  isPool: data[0].result,
  poolManagerLogic: data[1].result,
  isCustomCooldownDepositAllowed: data[2].result,
  customCooldown: data[3].result,
})

export const usePoolStatic = ({ address, chainId }: PoolContractCallParams) => {
  const result = useReadContracts({
    contracts: getContracts({ address, chainId }),
    query: {
      enabled: !!address && !isZeroAddress(address),
      staleTime: Infinity,
      select: selector,
    },
  })

  useContractReadErrorLogging({ error: result.error, status: result.status })

  return result
}
