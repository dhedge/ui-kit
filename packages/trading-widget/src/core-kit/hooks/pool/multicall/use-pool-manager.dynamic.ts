import { PoolLogicAbi, PoolManagerLogicAbi } from 'core-kit/abi'
import { AddressZero } from 'core-kit/const'
import { useManagerLogicAddress } from 'core-kit/hooks/pool/use-manager-logic-address'
import {
  useAccount,
  useContractReadErrorLogging,
  useReadContracts,
} from 'core-kit/hooks/web3'
import type {
  Address,
  MulticallReturnType,
  PoolContractCallParams,
} from 'core-kit/types'
import { isZeroAddress } from 'core-kit/utils'

type GetContractsParams = PoolContractCallParams & {
  managerLogicAddress: Address
  account: Address
}

const getContracts = ({
  address,
  chainId,
  managerLogicAddress,
  account,
}: GetContractsParams) =>
  [
    {
      address: managerLogicAddress,
      abi: PoolManagerLogicAbi,
      functionName: 'getFundComposition',
      chainId,
    },
    {
      address,
      abi: PoolLogicAbi,
      functionName: 'getExitRemainingCooldown',
      chainId,
      args: [account],
    },
  ] as const

type Data = MulticallReturnType<ReturnType<typeof getContracts>>

const selector = (data: Data) => ({
  getFundComposition: data[0].result,
  getExitRemainingCooldown: data[1].result,
})

export const usePoolManagerDynamic = ({
  address,
  chainId,
}: PoolContractCallParams) => {
  const { account = AddressZero } = useAccount()
  const managerLogicAddress = useManagerLogicAddress({ address, chainId })

  const result = useReadContracts({
    contracts: getContracts({
      address,
      chainId,
      managerLogicAddress: managerLogicAddress ?? AddressZero,
      account,
    }),
    query: {
      enabled: !!managerLogicAddress && !isZeroAddress(managerLogicAddress),
      select: selector,
    },
  })

  useContractReadErrorLogging({ error: result.error, status: result.status })

  return result
}
