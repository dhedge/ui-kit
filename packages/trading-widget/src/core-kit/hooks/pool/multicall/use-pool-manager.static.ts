import { PoolManagerLogicAbi } from 'core-kit/abi'
import { AddressZero } from 'core-kit/const'
import { useManagerLogicAddress } from 'core-kit/hooks/pool/use-manager-logic-address'
import {
  useContractReadErrorLogging,
  useReadContracts,
} from 'core-kit/hooks/web3'
import type {
  MulticallReturnType,
  PoolContractCallParams,
} from 'core-kit/types'
import { isZeroAddress } from 'core-kit/utils'

const getContracts = ({ address, chainId }: PoolContractCallParams) =>
  [
    {
      address: address,
      abi: PoolManagerLogicAbi,
      functionName: 'getFeeIncreaseInfo',
      chainId,
    },
    {
      address: address,
      abi: PoolManagerLogicAbi,
      functionName: 'minDepositUSD',
      chainId,
    },
  ] as const

type Data = MulticallReturnType<ReturnType<typeof getContracts>>

const selector = (data: Data) => ({
  getFeeIncreaseInfo: data[0].result,
  minDepositUSD: data[1].result,
})

export const usePoolManagerStatic = ({
  address,
  chainId,
}: PoolContractCallParams) => {
  const managerLogicAddress = useManagerLogicAddress({ address, chainId })

  const result = useReadContracts({
    contracts: getContracts({
      address: managerLogicAddress ?? AddressZero,
      chainId,
    }),
    query: {
      enabled: !!managerLogicAddress && !isZeroAddress(managerLogicAddress),
      staleTime: Infinity,
      select: selector,
    },
  })

  useContractReadErrorLogging({ error: result.error, status: result.status })

  return result
}
