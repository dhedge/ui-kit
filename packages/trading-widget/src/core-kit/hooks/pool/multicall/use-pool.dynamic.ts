import { PoolLogicAbi } from 'core-kit/abi'
import { DEFAULT_POLLING_INTERVAL } from 'core-kit/const'
import { useReadContracts } from 'core-kit/hooks/web3'
import type {
  MulticallReturnType,
  PoolContractCallParams,
} from 'core-kit/types'
import { isZeroAddress } from 'core-kit/utils'

const getContracts = ({ chainId, address }: PoolContractCallParams) =>
  [
    {
      address,
      abi: PoolLogicAbi,
      functionName: 'tokenPrice',
      chainId,
      args: [],
    },
    {
      address,
      abi: PoolLogicAbi,
      functionName: 'getFundSummary',
      chainId,
    },
  ] as const

type Data = MulticallReturnType<ReturnType<typeof getContracts>>

const selector = ([tokenPrice, getFundSummary]: Data) => {
  const summary = getFundSummary?.result

  return {
    tokenPrice: tokenPrice?.result?.toString(),
    totalValue: summary?.totalFundValue?.toString(),
    totalSupply: summary?.totalSupply?.toString(),
    isPrivateVault: summary?.privatePool,
    performanceFee: summary?.performanceFeeNumerator?.toString(),
    streamingFee: summary?.managerFeeNumerator?.toString(),
    entryFee: summary?.entryFeeNumerator?.toString(),
    exitFee: summary?.exitFeeNumerator?.toString(),
    managerAddress: summary?.manager?.toString(),
  }
}

export const usePoolDynamic = ({ address, chainId }: PoolContractCallParams) =>
  useReadContracts({
    contracts: getContracts({ address, chainId }),
    query: {
      enabled: !!address && !isZeroAddress(address),
      refetchInterval: DEFAULT_POLLING_INTERVAL,
      select: selector,
    },
  })
