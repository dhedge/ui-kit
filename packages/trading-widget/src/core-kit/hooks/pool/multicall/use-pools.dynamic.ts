import chunk from 'lodash.chunk'

import { PoolLogicAbi } from 'core-kit/abi'
import { AddressZero, DEFAULT_CHAIN_ID } from 'core-kit/const'
import { useTradingPanelPoolConfigs } from 'core-kit/hooks/state'
import {
  useAccount,
  useContractReadErrorLogging,
  useReadContracts,
} from 'core-kit/hooks/web3'
import type {
  Address,
  ContractFunctionReturnType,
  DynamicPoolContractData,
  PoolContractAccountCallParams,
} from 'core-kit/types'

const getPoolContracts = ({
  account,
  chainId,
  address,
}: PoolContractAccountCallParams) =>
  [
    {
      address,
      abi: PoolLogicAbi,
      functionName: 'balanceOf',
      chainId,
      args: [account ?? AddressZero],
    },
    {
      address,
      abi: PoolLogicAbi,
      functionName: 'tokenPrice',
      chainId,
      args: [],
    },
    {
      address,
      abi: PoolLogicAbi,
      functionName: 'getFundSummary',
      chainId,
    },
  ] as const

type GetFundSummary = ContractFunctionReturnType<
  typeof PoolLogicAbi,
  'view',
  'getFundSummary'
>

type PoolsMap = Record<Address, DynamicPoolContractData>

const getContracts = (pools: PoolContractAccountCallParams[]) =>
  pools.flatMap(getPoolContracts)

const POOL_CHUNK_SIZE = getPoolContracts({
  account: AddressZero,
  chainId: DEFAULT_CHAIN_ID,
  address: AddressZero,
}).length

export const usePoolsDynamic = () => {
  const { account = AddressZero } = useAccount()
  const pools = useTradingPanelPoolConfigs()

  const result = useReadContracts({
    contracts: getContracts(pools.map((pool) => ({ ...pool, account }))),
    query: {
      select: (data) =>
        chunk(data, POOL_CHUNK_SIZE).reduce<PoolsMap>(
          (acc, [balanceOf, tokenPrice, getFundSummary], index) => {
            const poolAddress = pools?.[index]?.address ?? AddressZero
            const summary = getFundSummary?.result as GetFundSummary

            return {
              ...acc,
              [poolAddress]: {
                userBalance: balanceOf?.result?.toString(),
                tokenPrice: tokenPrice?.result?.toString(),
                totalValue: summary?.totalFundValue?.toString(),
                totalSupply: summary?.totalSupply?.toString(),
                isPrivateVault: summary?.privatePool,
                performanceFee: summary?.performanceFeeNumerator?.toString(),
                streamingFee: summary?.managerFeeNumerator?.toString(),
                entryFee: summary?.entryFeeNumerator?.toString(),
              },
            }
          },
          {},
        ),
    },
  })

  useContractReadErrorLogging({ error: result.error, status: result.status })

  return result
}
