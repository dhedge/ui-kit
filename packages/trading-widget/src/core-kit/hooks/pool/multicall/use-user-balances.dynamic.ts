import chunk from 'lodash.chunk'

import type { UseReadContractsReturnType } from 'wagmi'
import { optimism } from 'wagmi/chains'

import { PoolLogicAbi } from 'core-kit/abi'
import { AddressZero, DEFAULT_POLLING_INTERVAL } from 'core-kit/const'
import { useTradingPanelPoolConfigs } from 'core-kit/hooks/state'
import { useNetwork, useReadContracts } from 'core-kit/hooks/web3'
import type { Address, PoolContractAccountCallParams } from 'core-kit/types'
import { isZeroAddress } from 'core-kit/utils'

type UseUserBalancesDynamicParams = {
  account: Address
}

type UserBalancesMap = Record<
  Address,
  {
    balance: string | undefined
    tokenPrice: string | undefined
  }
>

const getPoolContracts = ({
  account,
  chainId,
  address,
}: PoolContractAccountCallParams) =>
  [
    {
      address,
      abi: PoolLogicAbi,
      functionName: 'balanceOf',
      chainId,
      args: [account],
    },
    {
      address,
      abi: PoolLogicAbi,
      functionName: 'tokenPrice',
      chainId,
      args: [],
    },
  ] as const

const getContracts = (pools: PoolContractAccountCallParams[]) =>
  pools.flatMap(getPoolContracts)

const POOL_CHUNK_SIZE = getPoolContracts({
  account: AddressZero,
  chainId: optimism.id,
  address: AddressZero,
}).length

export const useUserBalancesDynamic = ({
  account,
}: UseUserBalancesDynamicParams): UseReadContractsReturnType<
  ReturnType<typeof getPoolContracts>,
  true,
  UserBalancesMap
> => {
  const { chainId } = useNetwork()
  const pools = useTradingPanelPoolConfigs()

  return useReadContracts({
    contracts: getContracts(pools.map((pool) => ({ ...pool, account }))),
    query: {
      select: (data) =>
        chunk(data, POOL_CHUNK_SIZE).reduce<UserBalancesMap>(
          (acc, [balanceOf, tokenPrice], index) => {
            const poolAddress = pools?.[index]?.address ?? AddressZero

            return {
              ...acc,
              [poolAddress]: {
                balance: balanceOf?.result?.toString(),
                tokenPrice: tokenPrice?.result?.toString(),
              },
            }
          },
          {},
        ),
      placeholderData: (previousValue, previousQuery) => {
        if (previousQuery?.queryKey?.[1].chainId !== chainId) {
          return previousValue
        }
      },
      refetchInterval: DEFAULT_POLLING_INTERVAL,
      enabled: !isZeroAddress(account),
    },
  })
}
