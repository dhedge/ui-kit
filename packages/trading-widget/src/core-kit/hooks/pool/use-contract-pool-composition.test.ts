import { expect } from 'vitest'

import { erc20Abi } from 'core-kit/abi'
import { DEFAULT_PRECISION, optimism } from 'core-kit/const'
import * as poolHooks from 'core-kit/hooks/pool'
import * as poolMulticallHooks from 'core-kit/hooks/pool/multicall'
import * as stateHooks from 'core-kit/hooks/state'
import * as web3Hooks from 'core-kit/hooks/web3'
import type { PoolComposition } from 'core-kit/types'
import { TEST_ADDRESS } from 'tests/mocks'
import { renderHook } from 'tests/test-utils'

import { useContractPoolComposition } from './use-contract-pool-composition'

vi.mock('core-kit/hooks/state', () => ({
  useTradingPanelPoolFallbackData: vi.fn(),
}))

vi.mock('core-kit/hooks/web3', () => ({
  useReadContracts: vi.fn(),
  useReadContract: vi.fn(),
  useContractReadsErrorLogging: vi.fn(),
}))
vi.mock('core-kit/hooks/pool', () => ({
  useSynthetixV3AssetBalance: vi.fn(),
  useManagerLogicAddress: vi.fn(),
}))

vi.mock('core-kit/hooks/pool/multicall', () => ({
  usePoolManagerDynamic: vi.fn(),
}))

describe('useContractPoolComposition', () => {
  it('should call usePoolManagerDynamic', () => {
    const address = TEST_ADDRESS
    const chainId = optimism.id
    const fallbackPoolComposition: PoolComposition = {
      tokenName: 'fallback_tokenName',
      rate: 'fallback_rate',
      amount: 'fallback_amount',
      isDeposit: true,
      tokenAddress: TEST_ADDRESS,
      precision: DEFAULT_PRECISION,
      asset: {
        iconSymbols: ['fallback_iconSymbol'],
      },
    }
    const fallbackPoolData = { poolCompositions: [fallbackPoolComposition] }
    const contractFundComposition = [
      [{ asset: TEST_ADDRESS, isDeposit: true }],
      [BigInt(1)],
      [BigInt(1)],
    ]
    const symbol = 'symbol'
    const decimals = DEFAULT_PRECISION
    const managerLogicAddress = TEST_ADDRESS

    vi.mocked(stateHooks.useTradingPanelPoolFallbackData).mockImplementation(
      () =>
        [fallbackPoolData, vi.fn()] as unknown as ReturnType<
          typeof stateHooks.useTradingPanelPoolFallbackData
        >,
    )
    vi.mocked(poolMulticallHooks.usePoolManagerDynamic).mockImplementationOnce(
      () =>
        ({
          data: { getFundComposition: contractFundComposition },
          isFetched: true,
        }) as unknown as ReturnType<
          typeof poolMulticallHooks.usePoolManagerDynamic
        >,
    )
    vi.mocked(web3Hooks.useReadContracts).mockImplementationOnce(
      () =>
        ({
          data: [{ result: symbol }, { result: decimals }],
        }) as unknown as ReturnType<typeof web3Hooks.useReadContracts>,
    )

    renderHook(() => useContractPoolComposition({ address, chainId }))

    expect(poolHooks.useSynthetixV3AssetBalance).toHaveBeenCalledWith({
      vaultAddress: address,
      chainId,
      disabled: true,
    })
    expect(poolMulticallHooks.usePoolManagerDynamic).toHaveBeenCalledTimes(1)
    expect(poolMulticallHooks.usePoolManagerDynamic).toHaveBeenCalledWith(
      expect.objectContaining({
        address: managerLogicAddress,
        chainId,
      }),
    )
  })

  it('should call symbol and decimals methods on erc20Abi for all composition assets', () => {
    const address = TEST_ADDRESS
    const chainId = optimism.id
    const fallbackPoolComposition: PoolComposition = {
      tokenName: 'fallback_tokenName',
      rate: 'fallback_rate',
      amount: 'fallback_amount',
      isDeposit: true,
      tokenAddress: TEST_ADDRESS,
      precision: DEFAULT_PRECISION,
      asset: {
        iconSymbols: ['fallback_iconSymbol'],
      },
    }
    const fallbackPoolData = { poolCompositions: [fallbackPoolComposition] }
    const contractFundComposition = [
      [{ asset: TEST_ADDRESS, isDeposit: true }],
      [BigInt(1)],
      [BigInt(1)],
    ]
    const symbol = 'symbol'
    const decimals = DEFAULT_PRECISION

    vi.mocked(stateHooks.useTradingPanelPoolFallbackData).mockImplementation(
      () =>
        [fallbackPoolData, vi.fn()] as unknown as ReturnType<
          typeof stateHooks.useTradingPanelPoolFallbackData
        >,
    )
    vi.mocked(poolMulticallHooks.usePoolManagerDynamic).mockImplementationOnce(
      () =>
        ({
          data: { getFundComposition: contractFundComposition },
          isFetched: true,
        }) as unknown as ReturnType<
          typeof poolMulticallHooks.usePoolManagerDynamic
        >,
    )
    vi.mocked(web3Hooks.useReadContracts).mockImplementationOnce(
      () =>
        ({
          data: [{ result: symbol }, { result: decimals }],
        }) as unknown as ReturnType<typeof web3Hooks.useReadContracts>,
    )

    renderHook(() => useContractPoolComposition({ address, chainId }))

    expect(web3Hooks.useReadContracts).toHaveBeenCalledTimes(1)
    expect(web3Hooks.useReadContracts).toHaveBeenCalledWith(
      expect.objectContaining({
        contracts: expect.arrayContaining([
          expect.objectContaining({
            address: TEST_ADDRESS,
            abi: erc20Abi,
            functionName: 'symbol',
            chainId,
          }),
          expect.objectContaining({
            address: TEST_ADDRESS,
            abi: erc20Abi,
            functionName: 'decimals',
            chainId,
          }),
        ]),
      }),
    )
  })

  it('should return extended fund composition data', () => {
    const address = TEST_ADDRESS
    const chainId = optimism.id
    const fallbackPoolComposition: PoolComposition = {
      tokenName: 'fallback_tokenName',
      rate: 'fallback_rate',
      amount: 'fallback_amount',
      isDeposit: true,
      tokenAddress: TEST_ADDRESS,
      precision: DEFAULT_PRECISION,
      asset: {
        iconSymbols: ['fallback_iconSymbol'],
      },
    }
    const fallbackPoolData = { poolCompositions: [fallbackPoolComposition] }
    const contractFundComposition = [
      [{ asset: TEST_ADDRESS, isDeposit: true }],
      [BigInt(1)],
      [BigInt(1)],
    ]
    const symbol = 'symbol'
    const decimals = DEFAULT_PRECISION

    vi.mocked(stateHooks.useTradingPanelPoolFallbackData).mockImplementation(
      () =>
        [fallbackPoolData, vi.fn()] as unknown as ReturnType<
          typeof stateHooks.useTradingPanelPoolFallbackData
        >,
    )
    vi.mocked(poolMulticallHooks.usePoolManagerDynamic).mockImplementationOnce(
      () =>
        ({
          data: { getFundComposition: contractFundComposition },
          isFetched: true,
        }) as unknown as ReturnType<
          typeof poolMulticallHooks.usePoolManagerDynamic
        >,
    )
    vi.mocked(web3Hooks.useReadContracts).mockImplementation(
      () =>
        ({
          data: [{ result: symbol }, { result: decimals }],
        }) as unknown as ReturnType<typeof web3Hooks.useReadContracts>,
    )

    const { result } = renderHook(() =>
      useContractPoolComposition({ address, chainId }),
    )

    expect(result.current).toMatchSnapshot()
  })

  // temporary disabled
  // it('should return extended fund composition data with synthetix v3 asset balance', () => {
  //   const address = DHEDGE_SYNTHETIX_V3_VAULT_ADDRESSES[0] as Address
  //   const chainId = optimism.id
  //   const fallbackPoolComposition: PoolComposition = {
  //     tokenName: 'fallback_tokenName',
  //     rate: 'fallback_rate',
  //     amount: 'fallback_amount',
  //     isDeposit: true,
  //     tokenAddress: TEST_ADDRESS,
  //     precision: DEFAULT_PRECISION,
  //     asset: {
  //       iconSymbols: ['fallback_iconSymbol'],
  //     },
  //   }
  //   const fallbackPoolData = { poolCompositions: [fallbackPoolComposition] }
  //   const contractFundComposition = [
  //     [
  //       { asset: TEST_ADDRESS, isDeposit: true },
  //       { asset: DHEDGE_SYNTHETIX_V3_ASSETS_MAP[chainId], isDeposit: false },
  //     ],
  //     [BigInt(1), BigInt(2)],
  //     [BigInt(1), BigInt(2)],
  //   ]
  //   const symbol = 'symbol'
  //   const decimals = DEFAULT_PRECISION
  //
  //   vi.mocked(stateHooks.useTradingPanelPoolFallbackData).mockImplementation(
  //     () =>
  //       [fallbackPoolData, vi.fn()] as unknown as ReturnType<
  //         typeof stateHooks.useTradingPanelPoolFallbackData
  //       >,
  //   )
  //   vi.mocked(poolHooks.useSynthetixV3AssetBalance).mockImplementation(
  //     () => '3',
  //   )
  //   vi.mocked(poolMulticallHooks.usePoolManagerDynamic).mockImplementationOnce(
  //     () =>
  //       ({
  //         data: { getFundComposition: contractFundComposition },
  //         isFetched: true,
  //       }) as unknown as ReturnType<
  //         typeof poolMulticallHooks.usePoolManagerDynamic
  //       >,
  //   )
  //   vi.mocked(web3Hooks.useReadContracts).mockImplementation(
  //     () =>
  //       ({
  //         data: [{ result: symbol }, { result: decimals }],
  //       }) as unknown as ReturnType<typeof web3Hooks.useReadContracts>,
  //   )
  //
  //   const { result } = renderHook(() =>
  //     useContractPoolComposition({ address, chainId }),
  //   )
  //
  //   expect(poolHooks.useSynthetixV3AssetBalance).toHaveBeenCalledWith({
  //     vaultAddress: address,
  //     chainId,
  //     disabled: false,
  //   })
  //   expect(result.current).toMatchSnapshot()
  // })
})
