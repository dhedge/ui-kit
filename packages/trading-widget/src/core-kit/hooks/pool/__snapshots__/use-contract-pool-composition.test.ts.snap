// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useContractPoolComposition > should return extended fund composition data 1`] = `
[
  {
    "amount": "1",
    "asset": {
      "iconSymbols": [
        "fallback_iconSymbol",
      ],
    },
    "isDeposit": true,
    "precision": 18,
    "rate": "1",
    "tokenAddress": "0xtest",
    "tokenName": "symbol",
  },
]
`;

exports[`useContractPoolComposition > should return extended fund composition data with synthetix v3 asset balance 1`] = `
[
  {
    "amount": "1",
    "asset": {
      "iconSymbols": [
        "fallback_iconSymbol",
      ],
    },
    "isDeposit": true,
    "precision": 18,
    "rate": "1",
    "tokenAddress": "0xtest",
    "tokenName": "symbol",
  },
  {
    "amount": "3",
    "asset": {
      "iconSymbols": [],
    },
    "isDeposit": false,
    "precision": 18,
    "rate": "2",
    "tokenAddress": "0xffffffaeff0b96ea8e4f94b2253f31abdd875847",
    "tokenName": "0xffff...5847",
  },
]
`;
